src/Foundatio/Caching/InMemoryCacheClient.cs

```diff
@@ -595,39 +616,38 @@ private async Task<bool> SetInternalAsync(string key, CacheEntry entry, bool add
 
         Interlocked.Increment(ref _writes);
 
+        bool wasUpdated = true;
+        bool isTraceLogLevelEnabled = _logger.IsEnabled(LogLevel.Trace);
         if (addOnly)
         {
-            if (!_memory.TryAdd(key, entry))
+            _memory.AddOrUpdate(key, entry, (_, existingEntry) =>
```

Comment:
Converted this to an AddOrUpdate as it can better handle concurrent attempts. Learned a few things from https://medium.com/gft-engineering/correctly-using-concurrentdictionarys-addorupdate-method-94b7b41719d6